Flaky: k8s.io/kubernetes/pkg/controller/framework_test.TestUpdate

```
E1109 11:31:44.493904    3385 util.go:82] Recovered from panic: "sync: negative WaitGroup counter" (sync: negative WaitGroup counter)
/home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/util.go:76
/home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/util.go:64
/usr/lib/golang/src/runtime/asm_amd64.s:437
/usr/lib/golang/src/runtime/panic.go:423
/usr/lib/golang/src/sync/waitgroup.go:70
/usr/lib/golang/src/sync/waitgroup.go:95
/home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/framework/controller_test.go:368
/home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/framework/controller.go:182
<autogenerated>:23
/home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/framework/controller.go:252
/home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/framework/controller.go:126
/home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/framework/controller.go:96
/home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/util.go:126
/home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/util.go:127
/home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/framework/controller.go:96
/usr/lib/golang/src/runtime/asm_amd64.s:1696
--- FAIL: TestUpdate (1.00s)
panic: sync: WaitGroup is reused before previous Wait has returned [recovered]
        panic: sync: WaitGroup is reused before previous Wait has returned

goroutine 7 [running]:
testing.tRunner.func1(0xc8201ae990)
        /usr/lib/golang/src/testing/testing.go:450 +0x171
sync.(*WaitGroup).Wait(0xc820230b80)
        /usr/lib/golang/src/sync/waitgroup.go:128 +0x114
k8s.io/kubernetes/pkg/controller/framework_test.TestUpdate(0xc8201ae990)
        /home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/framework/controller_test.go:393 +0xb11
testing.tRunner(0xc8201ae990, 0x1887d90)
        /usr/lib/golang/src/testing/testing.go:456 +0x98
created by testing.RunTests
        /usr/lib/golang/src/testing/testing.go:561 +0x86d

goroutine 1 [chan receive]:
testing.RunTests(0x12fc4c8, 0x1887d60, 0x3, 0x3, 0xc820230a01)
        /usr/lib/golang/src/testing/testing.go:562 +0x8ad
testing.(*M).Run(0xc8200d3ee8, 0xc82006f938)
        /usr/lib/golang/src/testing/testing.go:494 +0x70
main.main()
        k8s.io/kubernetes/pkg/controller/framework/_test/_testmain.go:64 +0x116

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
        /usr/lib/golang/src/runtime/asm_amd64.s:1696 +0x1

goroutine 5 [chan receive]:
github.com/golang/glog.(*loggingT).flushDaemon(0x18a6080)
        /home/vagrant/go/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/golang/glog/glog.go:879 +0x67
created by github.com/golang/glog.init.1
        /home/vagrant/go/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/golang/glog/glog.go:410 +0x297

goroutine 8 [chan receive]:
k8s.io/kubernetes/pkg/watch.(*Broadcaster).loop(0xc8201a4800)
        /home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/watch/mux.go:198 +0x58
created by k8s.io/kubernetes/pkg/watch.NewBroadcaster
        /home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/watch/mux.go:74 +0x128

goroutine 9 [select]:
k8s.io/kubernetes/pkg/util.Until(0xc8200cff68, 0x3b9aca00, 0xc8202401e0)
        /home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/util.go:128 +0x125
k8s.io/kubernetes/pkg/controller/framework.(*Controller).Run(0xc82023cee0, 0xc8202401e0)
        /home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/framework/controller.go:96 +0x1c2
created by k8s.io/kubernetes/pkg/controller/framework_test.TestUpdate
        /home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/framework/controller_test.go:377 +0x869

goroutine 55 [runnable]:
k8s.io/kubernetes/pkg/client/cache.(*Reflector).watchHandler(0xc8201aeab0, 0x7fa38f8852e8, 0xc820277cb0, 0xc82004dc88, 0xc820241ec0, 0xc8202401e0, 0x0, 0x0)
        /home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/cache/reflector.go:312 +0x13df
k8s.io/kubernetes/pkg/client/cache.(*Reflector).ListAndWatch(0xc8201aeab0, 0xc8202401e0, 0x0, 0x0)
        /home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/cache/reflector.go:279 +0x1215
k8s.io/kubernetes/pkg/client/cache.(*Reflector).RunUntil.func1()
        /home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/cache/reflector.go:164 +0x29
k8s.io/kubernetes/pkg/util.Until.func1(0xc820216ea0)
        /home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/util.go:126 +0x4b
k8s.io/kubernetes/pkg/util.Until(0xc820216ea0, 0x3b9aca00, 0xc8202401e0)
        /home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/util.go:127 +0x68
created by k8s.io/kubernetes/pkg/client/cache.(*Reflector).RunUntil
        /home/vagrant/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/cache/reflector.go:164 +0x9a
```

I got this by running the following against commit d65c61f:

```
export GOPATH=/home/vagrant/go/src/k8s.io/kubernetes/_output/local/go:/home/vagrant/go/src/k8s.io/kubernetes/Godeps/_workspace
go test -c k8s.io/kubernetes/pkg/controller/framework
while true; do GOMAXPROCS=4 ./framework.test -test.run TestUpdate; if [ $? != 0 ]; then break; fi; done
```

There have been previous attempts to fix this:

- #13058
- #14881

It appears it's still an issue :cry: 

We're also seeing it in OpenShift (https://github.com/openshift/origin/issues/5389), although our code hasn't pulled in the fix from #14881 quite yet.

--- a/pkg/controller/framework/controller_test.go
+++ b/pkg/controller/framework/controller_test.go
@@ -311,35 +311,42 @@ func TestUpdate(t *testing.T) {
 		pair{FROM, FROM}: true,
 	}
 
-	pod := func(name, check string) *api.Pod {
-		return &api.Pod{
+	pod := func(name, check string, final bool) *api.Pod {
+		p := &api.Pod{
 			ObjectMeta: api.ObjectMeta{
 				Name:   name,
 				Labels: map[string]string{"check": check},
 			},
 		}
+		if final {
+			p.Labels["final"] = "true"
+		}
+		return p
+	}
+	deletePod := func(p *api.Pod) bool {
+		return p.Labels["final"] == "true"
 	}
 
 	tests := []func(string){
 		func(name string) {
 			name = "a-" + name
-			source.Add(pod(name, FROM))
-			source.Modify(pod(name, TO))
+			source.Add(pod(name, FROM, false))
+			source.Modify(pod(name, TO, true))
 		},
 		func(name string) {
 			name = "b-" + name
-			source.Add(pod(name, FROM))
-			source.ModifyDropWatch(pod(name, TO))
+			source.Add(pod(name, FROM, false))
+			source.ModifyDropWatch(pod(name, TO, true))
 		},
 		func(name string) {
 			name = "c-" + name
-			source.AddDropWatch(pod(name, FROM))
-			source.Modify(pod(name, ADD_MISSED))
-			source.Modify(pod(name, TO))
+			source.AddDropWatch(pod(name, FROM, false))
+			source.Modify(pod(name, ADD_MISSED, false))
+			source.Modify(pod(name, TO, true))
 		},
 		func(name string) {
 			name = "d-" + name
-			source.Add(pod(name, FROM))
+			source.Add(pod(name, FROM, true))
 		},
 	}
 
@@ -362,7 +369,9 @@ func TestUpdate(t *testing.T) {
 				if !allowedTransitions[pair{from, to}] {
 					t.Errorf("observed transition %q -> %q for %v", from, to, n.Name)
 				}
-				source.Delete(n)
+				if deletePod(n) {
+					source.Delete(n)
+				}
 			},
 			DeleteFunc: func(obj interface{}) {
 				testDoneWG.Done()

